File: LLM-Translations/GPT-4/codenet/Python/Go/s233116196.go

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	input := strings.Split(scanner.Text(), " ")
	N, _ := strconv.Atoi(input[0])
	M, _ := strconv.Atoi(input[1])

	G := make(map[int][]int)
	for i := 1; i <= N; i++ {
		G[i] = []int{}
	}

	for i := 0; i < M; i++ {
		scanner.Scan()
		input := strings.Split(scanner.Text(), " ")
		a, _ := strconv.Atoi(input[0])
		b, _ := strconv.Atoi(input[1])
		G[a] = append(G[a], b)
		G[b] = append(G[b], a)
	}

	col := make([]int, N+1)
	cnt := 0
	for i := 1; i <= N; i++ {
		if col[i] < 0 {
			col[i] = cnt
			que := []int{i}
			for len(que) > 0 {
				x := que[0]
				que = que[1:]
				for _, y := range G[x] {
					if col[y] < 0 {
						col[y] = cnt
						que = append(que, y)
					}
				}
			}
			cnt++
		}
	}

	cmax := -1
	for _, v := range col {
		if v > cmax {
			cmax = v
		}
	}
	fmt.Println(cmax)
}

// 

File: LLM-Translations/GPT-4/codenet/Python/C/s233116196.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

int main() {
    int N, M;
    scanf("%d %d", &N, &M);

    int G[N+1][N+1];
    memset(G, 0, sizeof(G));

    for (int i = 0; i < M; i++) {
        int a, b;
        scanf("%d %d", &a, &b);
        G[a][b] = 1;
        G[b][a] = 1;
    }

    int col[N+1];
    for (int i = 0; i <= N; i++) {
        col[i] = -1;
    }

    int cnt = 0;
    for (int i = 1; i <= N; i++) {
        if (col[i] < 0) {
            col[i] = cnt;
            int que[N+1], head = 0, tail = 0;
            que[tail++] = i;

            while (head < tail) {
                int x = que[head++];
                for (int y = 1; y <= N; y++) {
                    if (G[x][y] && col[y] < 0) {
                        col[y] = cnt;
                        que[tail++] = y;
                    }
                }
            }

            cnt += 1;
        }
    }

    int cmax = col[0];
    for (int i = 1; i <= N; i++) {
        if (col[i] > cmax) {
            cmax = col[i];
        }
    }

    printf("%d\n", cmax);

    return 0;
}
